File: jolkr-app/src/utils/permissions.ts
/** Bitfield permission flags — mirrors backend jolkr_common::Permissions */

// General
export const ADMINISTRATOR = 1 << 0;
export const VIEW_CHANNELS = 1 << 1;
export const MANAGE_CHANNELS = 1 << 2;
export const MANAGE_ROLES = 1 << 3;
export const MANAGE_SERVER = 1 << 4;

// Membership
export const KICK_MEMBERS = 1 << 5;
export const BAN_MEMBERS = 1 << 6;
export const CREATE_INVITE = 1 << 7;
export const CHANGE_NICKNAME = 1 << 8;
export const MANAGE_NICKNAMES = 1 << 9;

// Text
export const SEND_MESSAGES = 1 << 10;
export const EMBED_LINKS = 1 << 11;
export const ATTACH_FILES = 1 << 12;
export const ADD_REACTIONS = 1 << 13;
export const MENTION_EVERYONE = 1 << 14;
export const MANAGE_MESSAGES = 1 << 15;
export const READ_MESSAGE_HISTORY = 1 << 16;

// Extended Moderation
export const MODERATE_MEMBERS = 1 << 28;
export const MANAGE_WEBHOOKS = 1 << 29;

// Voice
export const CONNECT = 1 << 20;
export const SPEAK = 1 << 21;
export const MUTE_MEMBERS = 1 << 23;
export const DEAFEN_MEMBERS = 1 << 24;
export const MOVE_MEMBERS = 1 << 25;

/** Check if a permissions bitfield has a specific permission. ADMINISTRATOR bypasses all. */
export function hasPermission(perms: number, permission: number): boolean {
  if ((perms & ADMINISTRATOR) !== 0) return true;
  return (perms & permission) === permission;
}

/** Channel-specific permission labels for channel overwrite editor */
export const CHANNEL_PERMISSION_LABELS: Array<{ key: string; flag: number; label: string; category: string }> = [
  { key: 'view_channels', flag: VIEW_CHANNELS, label: 'View Channel', category: 'General' },
  { key: 'manage_channels', flag: MANAGE_CHANNELS, label: 'Manage Channel', category: 'General' },
  { key: 'send_messages', flag: SEND_MESSAGES, label: 'Send Messages', category: 'Text' },
  { key: 'embed_links', flag: EMBED_LINKS, label: 'Embed Links', category: 'Text' },
  { key: 'attach_files', flag: ATTACH_FILES, label: 'Attach Files', category: 'Text' },
  { key: 'add_reactions', flag: ADD_REACTIONS, label: 'Add Reactions', category: 'Text' },
  { key: 'mention_everyone', flag: MENTION_EVERYONE, label: 'Mention Everyone', category: 'Text' },
  { key: 'manage_messages', flag: MANAGE_MESSAGES, label: 'Manage Messages', category: 'Text' },
  { key: 'read_message_history', flag: READ_MESSAGE_HISTORY, label: 'Read Message History', category: 'Text' },
  { key: 'connect', flag: CONNECT, label: 'Connect', category: 'Voice' },
  { key: 'speak', flag: SPEAK, label: 'Speak', category: 'Voice' },
  { key: 'mute_members', flag: MUTE_MEMBERS, label: 'Mute Members', category: 'Voice' },
  { key: 'deafen_members', flag: DEAFEN_MEMBERS, label: 'Deafen Members', category: 'Voice' },
  { key: 'move_members', flag: MOVE_MEMBERS, label: 'Move Members', category: 'Voice' },
];

/** Named permission labels for the role editor UI */
export const PERMISSION_LABELS: Array<{ key: string; flag: number; label: string; category: string }> = [
  { key: 'administrator', flag: ADMINISTRATOR, label: 'Administrator', category: 'General' },
  { key: 'manage_server', flag: MANAGE_SERVER, label: 'Manage Server', category: 'General' },
  { key: 'manage_channels', flag: MANAGE_CHANNELS, label: 'Manage Channels', category: 'General' },
  { key: 'manage_roles', flag: MANAGE_ROLES, label: 'Manage Roles', category: 'General' },
  { key: 'kick_members', flag: KICK_MEMBERS, label: 'Kick Members', category: 'Membership' },
  { key: 'ban_members', flag: BAN_MEMBERS, label: 'Ban Members', category: 'Membership' },
  { key: 'create_invite', flag: CREATE_INVITE, label: 'Create Invite', category: 'Membership' },
  { key: 'manage_nicknames', flag: MANAGE_NICKNAMES, label: 'Manage Nicknames', category: 'Membership' },
  { key: 'moderate_members', flag: MODERATE_MEMBERS, label: 'Timeout Members', category: 'Membership' },
  { key: 'manage_webhooks', flag: MANAGE_WEBHOOKS, label: 'Manage Webhooks', category: 'General' },
  { key: 'send_messages', flag: SEND_MESSAGES, label: 'Send Messages', category: 'Text' },
  { key: 'manage_messages', flag: MANAGE_MESSAGES, label: 'Manage Messages', category: 'Text' },
  { key: 'attach_files', flag: ATTACH_FILES, label: 'Attach Files', category: 'Text' },
  { key: 'add_reactions', flag: ADD_REACTIONS, label: 'Add Reactions', category: 'Text' },
  { key: 'mention_everyone', flag: MENTION_EVERYONE, label: 'Mention Everyone', category: 'Text' },
  { key: 'connect', flag: CONNECT, label: 'Connect', category: 'Voice' },
  { key: 'speak', flag: SPEAK, label: 'Speak', category: 'Voice' },
  { key: 'mute_members', flag: MUTE_MEMBERS, label: 'Mute Members', category: 'Voice' },
  { key: 'deafen_members', flag: DEAFEN_MEMBERS, label: 'Deafen Members', category: 'Voice' },
  { key: 'move_members', flag: MOVE_MEMBERS, label: 'Move Members', category: 'Voice' },
];

File: jolkr-server/crates/jolkr-common/src/permissions.rs
use serde::{Deserialize, Serialize};

/// Bitfield-based permissions, inspired by Discord's permission model.
///
/// Each permission is a single bit in a `u64`. Combine them with `|` and check
/// them with `has()`.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(transparent)]
pub struct Permissions(pub u64);

// ── Permission flag constants ──────────────────────────────────────────

impl Permissions {
    // General
    /// `NONE` constant.
    pub const NONE: u64 = 0;
    /// `ADMINISTRATOR` constant.
    pub const ADMINISTRATOR: u64 = 1 << 0;
    /// `VIEW_CHANNELS` constant.
    pub const VIEW_CHANNELS: u64 = 1 << 1;
    /// `MANAGE_CHANNELS` constant.
    pub const MANAGE_CHANNELS: u64 = 1 << 2;
    /// `MANAGE_ROLES` constant.
    pub const MANAGE_ROLES: u64 = 1 << 3;
    /// `MANAGE_SERVER` constant.
    pub const MANAGE_SERVER: u64 = 1 << 4;

    // Membership
    /// `KICK_MEMBERS` constant.
    pub const KICK_MEMBERS: u64 = 1 << 5;
    /// `BAN_MEMBERS` constant.
    pub const BAN_MEMBERS: u64 = 1 << 6;
    /// `CREATE_INVITE` constant.
    pub const CREATE_INVITE: u64 = 1 << 7;
    /// `CHANGE_NICKNAME` constant.
    pub const CHANGE_NICKNAME: u64 = 1 << 8;
    /// `MANAGE_NICKNAMES` constant.
    pub const MANAGE_NICKNAMES: u64 = 1 << 9;

    // Text
    /// `SEND_MESSAGES` constant.
    pub const SEND_MESSAGES: u64 = 1 << 10;
    /// `EMBED_LINKS` constant.
    pub const EMBED_LINKS: u64 = 1 << 11;
    /// `ATTACH_FILES` constant.
    pub const ATTACH_FILES: u64 = 1 << 12;
    /// `ADD_REACTIONS` constant.
    pub const ADD_REACTIONS: u64 = 1 << 13;
    /// `MENTION_EVERYONE` constant.
    pub const MENTION_EVERYONE: u64 = 1 << 14;
    /// `MANAGE_MESSAGES` constant.
    pub const MANAGE_MESSAGES: u64 = 1 << 15;
    /// `READ_MESSAGE_HISTORY` constant.
    pub const READ_MESSAGE_HISTORY: u64 = 1 << 16;
    /// `USE_EXTERNAL_EMOJIS` constant.
    pub const USE_EXTERNAL_EMOJIS: u64 = 1 << 17;
    /// `SEND_TTS_MESSAGES` constant.
    pub const SEND_TTS_MESSAGES: u64 = 1 << 18;

    // Voice
    /// `CONNECT` constant.
    pub const CONNECT: u64 = 1 << 20;
    /// `SPEAK` constant.
    pub const SPEAK: u64 = 1 << 21;
    /// `VIDEO` constant.
    pub const VIDEO: u64 = 1 << 22;
    /// `MUTE_MEMBERS` constant.
    pub const MUTE_MEMBERS: u64 = 1 << 23;
    /// `DEAFEN_MEMBERS` constant.
    pub const DEAFEN_MEMBERS: u64 = 1 << 24;
    /// `MOVE_MEMBERS` constant.
    pub const MOVE_MEMBERS: u64 = 1 << 25;
    /// `USE_VOICE_ACTIVITY` constant.
    pub const USE_VOICE_ACTIVITY: u64 = 1 << 26;
    /// `PRIORITY_SPEAKER` constant.
    pub const PRIORITY_SPEAKER: u64 = 1 << 27;

    // Moderation (extended)
    /// `MODERATE_MEMBERS` constant.
    pub const MODERATE_MEMBERS: u64 = 1 << 28;
    /// `MANAGE_WEBHOOKS` constant.
    pub const MANAGE_WEBHOOKS: u64 = 1 << 29;

    /// Default permissions granted to @everyone role.
    pub const DEFAULT: u64 = Self::VIEW_CHANNELS
        | Self::SEND_MESSAGES
        | Self::READ_MESSAGE_HISTORY
        | Self::EMBED_LINKS
        | Self::ATTACH_FILES
        | Self::ADD_REACTIONS
        | Self::USE_EXTERNAL_EMOJIS
        | Self::CONNECT
        | Self::SPEAK
        | Self::VIDEO
        | Self::USE_VOICE_ACTIVITY
        | Self::CHANGE_NICKNAME
        | Self::CREATE_INVITE;

    /// All permissions combined.
    pub const ALL: u64 = u64::MAX;
}

// ── Methods ────────────────────────────────────────────────────────────

impl Permissions {
    /// Create an empty permission set.
    #[must_use] 
    pub const fn empty() -> Self {
        Self(Self::NONE)
    }

    /// Create a permission set with the default @everyone permissions.
    #[must_use] 
    pub const fn default_everyone() -> Self {
        Self(Self::DEFAULT)
    }

    /// Check whether a specific permission flag (or flags) is/are set.
    /// If the user has ADMINISTRATOR, every check returns true.
    #[must_use] 
    pub const fn has(&self, permission: u64) -> bool {
        if self.0 & Self::ADMINISTRATOR != 0 {
            return true;
        }
        self.0 & permission == permission
    }

    /// Add one or more permission flags.
    pub const fn add(&mut self, permission: u64) {
        self.0 |= permission;
    }

    /// Remove one or more permission flags.
    pub const fn remove(&mut self, permission: u64) {
        self.0 &= !permission;
    }

    /// Merge another permission set into this one (union).
    pub const fn merge(&mut self, other: Self) {
        self.0 |= other.0;
    }

    /// Returns the raw `u64` value.
    #[must_use] 
    pub const fn bits(&self) -> u64 {
        self.0
    }
}

impl Default for Permissions {
    fn default() -> Self {
        Self::empty()
    }
}

impl From<u64> for Permissions {
    fn from(v: u64) -> Self {
        Self(v)
    }
}

impl From<i64> for Permissions {
    fn from(v: i64) -> Self {
        Self(v as u64)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_has_single_permission() {
        let perms = Permissions(Permissions::SEND_MESSAGES);
        assert!(perms.has(Permissions::SEND_MESSAGES));
        assert!(!perms.has(Permissions::MANAGE_CHANNELS));
    }

    #[test]
    fn test_administrator_bypasses_all() {
        let perms = Permissions(Permissions::ADMINISTRATOR);
        assert!(perms.has(Permissions::SEND_MESSAGES));
        assert!(perms.has(Permissions::BAN_MEMBERS));
        assert!(perms.has(Permissions::MANAGE_SERVER));
    }

    #[test]
    fn test_add_remove() {
        let mut perms = Permissions::empty();
        perms.add(Permissions::SEND_MESSAGES);
        assert!(perms.has(Permissions::SEND_MESSAGES));

        perms.remove(Permissions::SEND_MESSAGES);
        assert!(!perms.has(Permissions::SEND_MESSAGES));
    }

    #[test]
    fn test_merge() {
        let mut a = Permissions(Permissions::SEND_MESSAGES);
        let b = Permissions(Permissions::MANAGE_CHANNELS);
        a.merge(b);
        assert!(a.has(Permissions::SEND_MESSAGES));
        assert!(a.has(Permissions::MANAGE_CHANNELS));
    }
}
